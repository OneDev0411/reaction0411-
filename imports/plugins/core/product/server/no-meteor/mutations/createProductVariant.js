import SimpleSchema from "simpl-schema";
import Logger from "@reactioncommerce/logger";
import Random from "@reactioncommerce/random";
import ReactionError from "@reactioncommerce/reaction-error";
import createProductOrVariant from "../utils/createProductOrVariant";
import isAncestorDeleted from "../utils/isAncestorDeleted";

const inputSchema = new SimpleSchema({
  parentId: String
});

/**
 * @method createProductVariant
 * @summary creates an empty variant on the product supplied
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Input arguments for the bulk operation
 * @param {String} input.parentId - the product or variant ID which we create new variant on
 * @return {String} created variantId
 */
export default async function createProductVariant(context, input) {
  inputSchema.validate(input);
  const { parentId } = input;
  const { collections, userHasPermission } = context;
  const { Products } = collections;

  // See that parent product exists
  const parentProduct = await Products.findOne({ _id: parentId });

  if (!parentProduct) {
    throw new ReactionError("not-found", "Product not found");
  }

  let product;
  let parentVariant;
  if (parentProduct.type === "variant") {
    product = await Products.findOne({ _id: parentProduct.ancestors[0] });
    parentVariant = parentProduct;
  } else {
    product = parentProduct;
    parentVariant = null;
  }

  // See that user has permission to create variant
  if (!userHasPermission(["createProduct", "product/admin", "product/create"], product.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  // Verify that parent is not deleted
  // Variants cannot be created on a deleted product
  if (await isAncestorDeleted(context, product, true)) {
    throw new ReactionError("server-error", "Unable to create product variant on a deleted product");
  }

  // get ancestors to build new ancestors array
  const { ancestors } = parentProduct;
  Array.isArray(ancestors) && ancestors.push(parentId);

  const newVariantId = Random.id();
  const newVariant = {
    _id: newVariantId,
    ancestors,
    shopId: product.shopId,
    type: "variant"
  };

  const isOption = ancestors.length > 1;

  const createdVariantId = await createProductOrVariant(context, newVariant, { product, parentVariant, isOption });

  if (!createdVariantId) {
    throw new ReactionError("server-error", "Unable to create product variant");
  }

  Logger.debug(`createProductVariant: created variant: ${createdVariantId} for ${parentId}`);

  const createdVariant = await Products.findOne({ _id: createdVariantId });

  if (!createdVariant) {
    throw new ReactionError("server-error", "Unable to retrieve newly created product variant");
  }

  return createdVariant;
}
