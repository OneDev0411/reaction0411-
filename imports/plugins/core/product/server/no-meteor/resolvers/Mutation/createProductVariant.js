import { decodeProductOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/product";
import { decodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";

/**
 *
 * @method createProductVariant
 * @summary initializes empty variant template
 * @param {Object} _ - unused
 * @param {Object} args - The input arguments
 * @param {Object} args.input - mutation input object
 * @param {String} args.input.clientMutationId - The mutation id
 * @param {String[]} args.input.productId - the product or variant ID which we create new variant on
 * @param {String[]} args.input.shopId - the shop to create the variant for
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} createProductVariant payload
 */
export default async function createProductVariant(_, { input }, context) {
  const {
    clientMutationId = null,
    productId,
    shopId
  } = input;

  const variant = await context.mutations.createProductVariant(context, {
    productId: decodeProductOpaqueId(productId),
    shopId: decodeShopOpaqueId(shopId)
  });

  return {
    clientMutationId,
    variant
  };
}
