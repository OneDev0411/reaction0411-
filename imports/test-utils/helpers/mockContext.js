const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
