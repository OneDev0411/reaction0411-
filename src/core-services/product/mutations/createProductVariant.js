import SimpleSchema from "simpl-schema";
import Logger from "@reactioncommerce/logger";
import Random from "@reactioncommerce/random";
import ReactionError from "@reactioncommerce/reaction-error";
import { ProductVariant } from "../simpleSchemas.js";
import isAncestorDeleted from "../utils/isAncestorDeleted.js";

const inputSchema = new SimpleSchema({
  productId: String,
  shopId: String
});

/**
 * @method createProductVariant
 * @summary creates an empty variant on the product supplied
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Input arguments for the bulk operation
 * @param {String} input.productId - the product or variant ID which we create new variant on
 * @param {String} input.shopId - the shop to create the variant for
 * @return {String} created variantId
 */
export default async function createProductVariant(context, input) {
  inputSchema.validate(input);
  const { collections } = context;
  const { Products } = collections;
  const { productId, shopId } = input;

  // See that user has permission to create variant
  await context.validatePermissions("reaction:legacy:products", "create", { shopId });

  // See that parent product exists
  const parentProduct = await Products.findOne({ _id: productId, shopId });

  if (!parentProduct) {
    throw new ReactionError("not-found", "Product not found");
  }

  let product;
  let parentVariant;
  if (parentProduct.type === "variant") {
    product = await Products.findOne({ _id: parentProduct.ancestors[0], shopId });
    parentVariant = parentProduct;
  } else {
    product = parentProduct;
    parentVariant = null;
  }

  // Verify that parent is not deleted
  // Variants cannot be created on a deleted product
  if (await isAncestorDeleted(context, product, true)) {
    throw new ReactionError("server-error", "Unable to create product variant on a deleted product");
  }

  // get ancestors to build new ancestors array
  let { ancestors } = parentProduct;
  if (Array.isArray(ancestors)) {
    ancestors.push(productId);
  } else {
    ancestors = [productId];
  }

  const newVariantId = Random.id();
  const createdAt = new Date();
  const newVariant = {
    _id: newVariantId,
    ancestors,
    createdAt,
    isDeleted: false,
    isVisible: false,
    shopId,
    type: "variant",
    updatedAt: createdAt,
    workflow: {
      status: "new"
    }
  };

  const isOption = ancestors.length > 1;

  // Apply custom transformations from plugins.
  for (const customFunc of context.getFunctionsOfType("mutateNewVariantBeforeCreate")) {
    // Functions of type "mutateNewVariantBeforeCreate" are expected to mutate the provided variant.
    // We need to run each of these functions in a series, rather than in parallel, because
    // we are mutating the same object on each pass.
    // eslint-disable-next-line no-await-in-loop
    await customFunc(newVariant, { context, isOption, parentVariant, product });
  }

  ProductVariant.validate(newVariant);

  await Products.insertOne(newVariant);

  Logger.debug(`createProductVariant: created variant: ${newVariantId} for ${productId}`);

  return newVariant;
}
