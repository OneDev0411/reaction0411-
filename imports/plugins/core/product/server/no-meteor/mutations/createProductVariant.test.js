import mockContext from "/imports/test-utils/helpers/mockContext";
import createProductVariant from "./createProductVariant";

mockContext.mutations.createProductVariant = jest.fn().mockName("mutations.createProductVariant");

test("throws if permission check fails", async () => {
  mockContext.collections.Products.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "PRODUCT_ID",
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(createProductVariant(mockContext, {
    parentId: "PRODUCT_ID"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["createProduct", "product/admin", "product/create"], "SHOP_ID");
});

test("throws if the parentId isn't supplied", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(createProductVariant(mockContext, {
    parentId: undefined
  })).rejects.toThrowErrorMatchingSnapshot();
});
