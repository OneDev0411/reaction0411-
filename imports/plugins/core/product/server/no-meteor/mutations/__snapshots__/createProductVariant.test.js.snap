// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the parentId isn't supplied 1`] = `"Parent ID is required"`;
